### ISSUE TITLE
Update dowload url

### ISSUE THREAD

**Issue by maulik9898** [tonyofrancis/Fetch#406]
Is it possible to update the URL of the current downloading file?
Considering the situation where some link expires after some time so to update our download with a new fresh link. this will save internet data as the previously downloaded file will be reused instead of creating a new file and starting the download again.

**Comment by tonyofrancis** [comment]
@maulik9898 You can go about this two different ways.
1. Use the `fetch.updateRequest` method. Wait till the fetch listener `onError` method  is called about the download failing due to url expiry. Then call updateRequest method. It will continue where the download left off.
```java
        fetch.addListener(new AbstractFetchListener() {
            @Override
            public void onError(@NotNull Download download, @NotNull Error error, @Nullable Throwable throwable) {
                super.onError(download, error, throwable);
                // url error.
                String newUrl = "test";
                Request request = new Request(newUrl, download.getFile());
                //update other request fields.
                fetch.updateRequest(download.getId(), request, true, new Func<Download>() {
                    @Override
                    public void call(@NotNull Download result) {
                        //updated/
                    }
                }, new Func<Error>() {
                    @Override
                    public void call(@NotNull Error result) {
                        //failed to update
                    }
                });
            }
        });
```
2. The second option requires you to create an custom downloader.
```java
     final FetchConfiguration fetchConfiguration = new FetchConfiguration.Builder(this)
                .enableRetryOnNetworkGain(true)
                .setDownloadConcurrentLimit(3)
                .setHttpDownloader(new MyCustomDownloader())
                .build();
Fetch fetch = Fetch.Impl.getInstance(fetchConfiguration);
    public static class MyCustomDownloader extends OkHttpDownloader {

        @NotNull
        @Override
        public Request onPreClientExecute(@NotNull OkHttpClient client, @NotNull Downloader.ServerRequest request) {
            final String url = request.getUrl();
            //check if url is expired via the client. then return a new request
            return new Request()
                    .newBuilder()
                    .url("new url if expired")
                    .addHeader("add headers", "add headers")
                    .build();
        }
    }
```
You can also override the CustomDownloaders `Response execute(request, interruptMonitor)` method and pass the super a new request with the new url.
```java
     public Response execute(@NotNull Downloader.ServerRequest request, @NotNull InterruptMonitor interruptMonitor) {
            final String url = request.getUrl();
            OkHttpClient client = getClient();
            //check if url is expired via the client. then return a new request
            Downloader.ServerRequest newRequest = new Downloader.ServerRequest("pass in the different params");
            return super.execute(newRequest, interruptMonitor);
        }

``` 
Please let me know if this helps.

**Comment by maulik9898** [comment]
@tonyofrancis  ok i will look into it.
There is one more issue how can i change concurrent download limit of fetch at runtime like on user pref change .?

**Comment by tonyofrancis** [comment]
@maulik9898  `fetch.setDownloadConcurrentLimit(999)`

**Comment by maulik9898** [comment]
@tonyofrancis for some reason using 
```
fetch.setDownloadConcurrentLimit(999) 
```
stops all running download i have to restart all again

**Comment by maulik9898** [comment]
fetch.updateRequest  does not resume the download from the previous position it starts download from the beginning and also results of fetch.updateRequest  will be updated download with new id so there will be two downloads of the same filename and different id as fetch only cancel the old download and not remove it. it also causes notification bug as there will be two downloads one new and one old.

here is log

```
2019-06-28 08:41:17.666 26073-26073/com.maulik.edm.free D/SERVICE PROGRESS: DownloadInfo(id=-1540025707, namespace='LibGlobalFetchLib', url='https://doc-08-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/2r29e6p9v20bj398k29k526s64ftknm0/1561687200000/12024208594662412395/*/1vYLRXeDvApNQq-k0rmlCrOcPN34e-LI5?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=166074811, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691439764, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=88000, downloadedBytesPerSecond=3011062)
2019-06-28 08:41:18.456 26073-26099/com.maulik.edm.free D/OpenGLRenderer: endAllActiveAnimators on 0x716ac1db00 (RippleDrawable) with handle 0x716ad34dc0
2019-06-28 08:41:18.664 26073-26073/com.maulik.edm.free D/SERVICE PROGRESS: DownloadInfo(id=-1540025707, namespace='LibGlobalFetchLib', url='https://doc-08-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/2r29e6p9v20bj398k29k526s64ftknm0/1561687200000/12024208594662412395/*/1vYLRXeDvApNQq-k0rmlCrOcPN34e-LI5?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=168925617, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691439764, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=92000, downloadedBytesPerSecond=2850016)
2019-06-28 08:41:19.663 26073-26073/com.maulik.edm.free D/SERVICE PROGRESS: DownloadInfo(id=-1540025707, namespace='LibGlobalFetchLib', url='https://doc-08-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/2r29e6p9v20bj398k29k526s64ftknm0/1561687200000/12024208594662412395/*/1vYLRXeDvApNQq-k0rmlCrOcPN34e-LI5?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=172423572, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691439764, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=87000, downloadedBytesPerSecond=2964760)
2019-06-28 08:41:20.424 26073-26100/com.maulik.edm.free D/LibGlobalFetchLib: DownloadManager cancelled download DownloadInfo(id=-1540025707, namespace='LibGlobalFetchLib', url='https://doc-08-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/2r29e6p9v20bj398k29k526s64ftknm0/1561687200000/12024208594662412395/*/1vYLRXeDvApNQq-k0rmlCrOcPN34e-LI5?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=173513107, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691439764, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=87000, downloadedBytesPerSecond=2964760)
2019-06-28 08:41:20.502 26073-26100/com.maulik.edm.free D/LibGlobalFetchLib: UpdatedRequest with id: -1540025707 to DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=172423572, total=430141753, status=QUEUED, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=-1, downloadedBytesPerSecond=-1)
2019-06-28 08:41:20.503 26073-26099/com.maulik.edm.free D/OpenGLRenderer: endAllActiveAnimators on 0x716e104800 (RippleDrawable) with handle 0x716feaaac0
2019-06-28 08:41:20.519 26073-26073/com.maulik.edm.free D/LibGlobalFetchLib: PriorityIterator backoffTime reset to 500 milliseconds
2019-06-28 08:41:20.682 26073-26073/com.maulik.edm.free D/LibGlobalFetchLib: PriorityIterator backoffTime reset to 500 milliseconds
2019-06-28 08:41:21.846 26073-26142/com.maulik.edm.free D/LibGlobalFetchLib: DownloadManager starting download DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=172423572, total=430141753, status=QUEUED, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=-1, downloadedBytesPerSecond=-1)
2019-06-28 08:41:23.249 26073-26100/com.maulik.edm.free D/LibGlobalFetchLib: PriorityIterator backoffTime increased to 1 minute(s)
2019-06-28 08:41:24.200 26073-26073/com.maulik.edm.free D/SERVICE PROGRESS: DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=1223491, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=351000, downloadedBytesPerSecond=1223491)
2019-06-28 08:41:25.203 26073-26073/com.maulik.edm.free D/SERVICE PROGRESS: DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=2280268, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=385000, downloadedBytesPerSecond=1112349)
2019-06-28 08:41:26.220 26073-26073/com.maulik.edm.free D/SERVICE PROGRESS: DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=3754806, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, 
2019-06-28 08:41:27.216 26073-26073/com.maulik.edm.free D/SERVICE PROGRESS: DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=8686361, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=154000, downloadedBytesPerSecond=2748683)
2019-06-28 08:41:27.958 26073-26082/com.maulik.edm.free W/System: A resource failed to call close. 
2019-06-28 08:41:27.958 26073-26082/com.maulik.edm.free I/chatty: uid=10123(com.maulik.edm.free) FinalizerDaemon identical 4 lines
2019-06-28 08:41:27.958 26073-26082/com.maulik.edm.free W/System: A resource failed to call close. 
2019-06-28 08:41:28.219 26073-26073/com.maulik.edm.free D/SERVICE PROGRESS: DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=14011150, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=116000, downloadedBytesPerSecond=3607385)


```

as you see fetch first cancel download then update request now as URL is new so id of the new download will be different than old one. After updating request fetch start to download but it is not able to resume from old point ( ***Link is resumable, I have tested this update behaviour on ADM app with same two links and it was able to update download without problem*** ).

**In summary, there is two issue**
1) The updated download does not resume from the previous point
2) Notification of old download is still shown and is also not removable

**Comment by maulik9898** [comment]
After some digging, I found that u delete old request from the database ( `fetchDatabaseManagerWrapper.delete(oldDownloadInfo)` ) but u do not update listener about removing a request that's why we see two notification as to the previous one never removed.

```

   override fun updateRequest(requestId: Int, newRequest: Request): Pair<Download, Boolean> {
        var oldDownloadInfo = fetchDatabaseManagerWrapper.get(requestId)
        if (oldDownloadInfo != null) {
            cancelDownloadsIfDownloading(listOf(oldDownloadInfo))
            oldDownloadInfo = fetchDatabaseManagerWrapper.get(requestId)
        }
        return if (oldDownloadInfo != null) {
            if (newRequest.file == oldDownloadInfo.file) {
                val newDownloadInfo = newRequest.toDownloadInfo()
                newDownloadInfo.namespace = namespace
                newDownloadInfo.downloaded = oldDownloadInfo.downloaded
                newDownloadInfo.total = oldDownloadInfo.total
                if (oldDownloadInfo.status == Status.DOWNLOADING) {
                    newDownloadInfo.status = Status.QUEUED
                    newDownloadInfo.error = defaultNoError
                } else {
                    newDownloadInfo.status = oldDownloadInfo.status
                    newDownloadInfo.error = oldDownloadInfo.error
                }
                fetchDatabaseManagerWrapper.delete(oldDownloadInfo)
                fetchDatabaseManagerWrapper.insert(newDownloadInfo)
                startPriorityQueueIfNotStarted()
                return Pair(newDownloadInfo, true)
            } else {
                delete(listOf(requestId))
                val enqueuePair = enqueue(newRequest)
                Pair(enqueuePair.first, enqueuePair.second == Error.NONE)
            }
        } else {
            throw FetchException(REQUEST_DOES_NOT_EXIST)
        }
    }
```

**Comment by tonyofrancis** [comment]
@maulik9898 

1. `setDownloadConcurrentLimit(999)` performs the correct behavior by stopping all current download and restarting them after the concurrent limit is set. Fetch downloads request by priority so it makes sense to restart the downloads. 
2. `updateRequest()` as the method names suggests it updates a request. Which means it has to delete the existing request before it sets the current one in the database. It would be a wrong behavior to call the remove callback for a request that is being `updated` and not removed.

I am currently working on updating the notification manager to provide better updates.

**Comment by maulik9898** [comment]
I was talking about the old request. as you first cancel old download then remove it from database and update database with the new request which has new id and it is not equal to the old one. So I think User should get notified that Old Download with download id xxx is changed to new download with download id yyy so that we can update UI accordingly.
What currently happens is that we get the progress of new Download but we don't know what happened to an old request.

```
2019-06-28 08:41:20.424 26073-26100/com.maulik.edm.free D/LibGlobalFetchLib: DownloadManager cancelled download DownloadInfo(id=-1540025707, namespace='LibGlobalFetchLib', url='https://doc-08-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/2r29e6p9v20bj398k29k526s64ftknm0/1561687200000/12024208594662412395/*/1vYLRXeDvApNQq-k0rmlCrOcPN34e-LI5?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=173513107, total=430141753, status=DOWNLOADING, error=NONE, networkType=ALL, created=1561691439764, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=87000, downloadedBytesPerSecond=2964760)
2019-06-28 08:41:20.502 26073-26100/com.maulik.edm.free D/LibGlobalFetchLib: UpdatedRequest with id: -1540025707 to DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=172423572, total=430141753, status=QUEUED, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=-1, downloadedBytesPerSecond=-1)
2019-06-28 08:41:20.503 26073-26099/com.maulik.edm.free D/OpenGLRenderer: endAllActiveAnimators on 0x716e104800 (RippleDrawable) with handle 0x716feaaac0
2019-06-28 08:41:20.519 26073-26073/com.maulik.edm.free D/LibGlobalFetchLib: PriorityIterator backoffTime reset to 500 milliseconds
2019-06-28 08:41:20.682 26073-26073/com.maulik.edm.free D/LibGlobalFetchLib: PriorityIterator backoffTime reset to 500 milliseconds
2019-06-28 08:41:21.846 26073-26142/com.maulik.edm.free D/LibGlobalFetchLib: DownloadManager starting download DownloadInfo(id=-1826123860, namespace='LibGlobalFetchLib', url='https://doc-0s-60-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/bvk6g1r2qlqho25h6vq45ao1gdac3hpq/1561687200000/12024208594662412395/*/1nsbEByVPgYB-8i3Hx2RVT_MIvIgKnFBW?e=download', file='/storage/emulated/0/Download/EDM/_igetintopc.com_Proteus_8.7_SP3_Professional_3.zip', group=500, priority=HIGH, headers={}, downloaded=172423572, total=430141753, status=QUEUED, error=NONE, networkType=ALL, created=1561691480427, tag=null, enqueueAction=UPDATE_ACCORDINGLY, identifier=0, downloadOnEnqueue=true, extras={"thread":"5"}, autoRetryMaxAttempts=0, autoRetryAttempts=0, etaInMilliSeconds=-1, downloadedBytesPerSecond=-1)

```

as you can see we get an update of new Download of id =-1826123860 but no info about the old request with id=-1540025707.

one solution would be to add `onUpdate(Download oldDownload, Download updatedDownload)` method in the listener to update about updating request

**Comment by tonyofrancis** [comment]
@maulik9898 this fix will be in the next release

**Comment by tonyofrancis** [comment]
@maulik9898  This issue is now fixed in new Fetch version 3.0.10/Androidx 3.1.4. The FetchListener onDelete method is now called first when updating a request.

**Comment by robbyl** [comment]
Having the same issue.
Updating the url makes the download to start from the beginning it does not continue from where I left

**Comment by W7orld** [comment]
Does we should calling `fetch.retry(result.getId()); `on update success?
```
  @Override
  public void call(@NotNull Download result) {
       //updated/
    fetch.retry(result.getId()); 
}
```
